Aarch64: fix narrow integer-register extension with Baldrdash ABI.

cfallin opened the pull request:
In the Baldrdash (SpiderMonkey) embedding, we must take care to
zero-extend all function arguments to callees in integer registers when
the types are narrower than 64 bits. This is because, unlike the native
SysV ABI, the Baldrdash ABI expects high bits to be cleared. Not doing
so leads to difficult-to-trace errors where high bits falsely tag an
int32 as e.g. an object pointer, leading to potential security issues.

<!--

Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

bjorn3 commented:
`AbiParam` has the `extension` field for this. It seems that it is not implemented AArch64 though. Maybe implement it and if `extension` is `None` and call conv extends baldrdash then use `Uext` instead.

cfallin commented:
I had forgotten about the `ArgumentExtension` value; we even used it for return values. Thanks for the reminder -- this is much better. Updated (and cleaned up the return-value side as well, now that we carry the extension type in the new-backend ABI signature).

julian-seward1 commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -1469,12 +1484,49 @@ impl ABICall for AArch64ABICall {
+                ));
+            }
+            &ABIArg::Stack(off, ty, ext) => {
+                if ext != ir::ArgumentExtension::None && ty_bits(ty) < 64 {
```
Presumably (correct me if I'm wrong) this is correct but a bit suboptimal, in that we're extending a value that we read from the stack in a previous insn; and in the ideal case we could have just done a suitable widening load?  Is that a correct understanding?  I'm not saying this should be fixed, but just checking I have this right.

julian-seward1 replied:
And ditto observation for the x64 equivalent of this, AFAICS.

cfallin replied:
Other way around -- we're storing a value in a register to the stack; sadly there's no store instruction (as far as I know) that can sign/zero-extend the bottom N bits of a GPR into 64 bits and then store that 64 bits, all in one go.

julian-seward1 commented on cranelift/filetests/filetests/vcode/aarch64/call.clif:
```diff
@@ -16,3 +16,32 @@ block0(v0: i64):
 ; nextln:  mov sp, fp
 ; nextln:  ldp fp, lr, [sp], #16
 ; nextln:  ret
+
```
I'm not 100% sure what I'm looking at here, but I was a bit mystified in that I can't see any signed widening insns in the expected outputs.  But it looks like the patch above handles both signed and unsigned widening.  If a signed-widening version of this test case does make sense, please can you add it?

bnjbvr replied:
+1; this only tests uext, not sext, which would be nice to have tests for!

cfallin replied:
Added `sext` tests as well; thanks!
